inspirehep/modules/crossref/ext.py

```diff
@@ -0,0 +1,146 @@
+
+import requests
+from urlparse import urljoin
+
+from flask import current_app, request, jsonify
+from flask_login import login_required
+
+response_code = {'success': 200,
+                 'notfound': 404,
+                 'malformed': 422,
+                 'multiplefound': 300}
+
+
+class CrossRef(object):
+
+    """CrossRef extension implementation.
+
+    Initialization of the extension:
+
+    >>> from flask import Flask
+    >>> from invenio_ext.crossref import CrossRef
+    >>> app = Flask("myapp")
+    >>> ext = CrossRef(app=app)
+
+    or alternatively using the factory pattern:
+
+    >>> app = Flask("myapp")
+    >>> ext = CrossRef()
+    >>> ext.init_app(app)
+    """
+
+    def __init__(self, app=None):
+        """Initialize extension."""
+        self.app = app
+        if app is not None:
+            self.init_app(app)
+
+    def init_app(self, app):
+        """Initialize a Flask application."""
+        app.config.setdefault("CROSSREF_API_URL",
+                              "http://api.crossref.org/works/")
+        app.config.setdefault("CROSSREF_ENDPOINT", "_doi.search")
+        app.config.setdefault("CROSSREF_URL_RULE", "/doi/search")
+
+        # Follow the Flask guidelines on usage of app.extensions
+        if not hasattr(app, "extensions"):
+            app.extensions = {}
+        if "crossref" in app.extensions:
+            raise RuntimeError("Flask application already initialized")
+        app.extensions["crossref"] = self
+
+        if app.config["CROSSREF_ENDPOINT"]:
+            app.add_url_rule(app.config["CROSSREF_URL_RULE"],
+                             app.config["CROSSREF_ENDPOINT"],
+                             login_required(self.search))
+
+    def get_response(self, crossref_doi):
+        """Get CrossRef response from the ``CROSSREF_API_URL`` page."""
+        response = requests.get(
+            urljoin(
+                current_app.config["CROSSREF_API_URL"],
+                "{term}".format(term=crossref_doi.strip()),
+            ),
+        )
+        return response
+
+    def get_json(self, doi):
+        """Get CrossRef json data."""
+        response = self.get_response(doi)
+        data, query = {}, {}
+
+        if response.status_code == 200:
+            if 'message' in response.json():
+                query = response.json().get('message')
+            else:
+                query = response.json()
+            data["status"] = "success"
+        elif response.status_code == 404:
+            data["status"] = "notfound"
+
+        data["source"] = "crossref"
+        data["query"] = query
+
+        return data
+
+    def search(self, doi=None):
+        """Search for given DOI."""
+        doi = doi or request.args.get("doi")
+
+        # query crossref
+        result = self.get_json(doi)
+
+        resp = jsonify(result)
+        resp.status_code = response_code.get(result['status'], 200)
+        return resp
+
+__all__ = ("CrossRef", )
```

Comment:
Yes, this two extensions used to be part of `invenio-ext`. If they are not needed there we can move them to `inspirehep`, the arXiv one had some tests ready https://github.com/inveniosoftware/invenio-ext/blob/master/tests/test_ext_arxiv.py